### PR TITLE
Batch2 cleanup

### PR DESCRIPTION
**Pull request by michaelabuckley**
Review fixes and cleanup from #4647

- Deleted methods in IJobPersistence unused in prod paths, and replaced their usage in tests.
- Lots of docs.
- Replace copy-on-write pattern in JobDefinitionRegistry with simple ConcurrentHashMap
- Fixed bad mappings from job ERRORED state to external APIs.  People think ERRORED is a failed state. `¯\_(ツ)_/¯`
- Added some spec tests for chunk purging
- Deprecated ERRORED.  For deleting in 6.8
- Lots of plans for 6.8. Too risky for 6.6

### CODE REVIEW

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobDefinitionRegistry.java [hunk 1]
```diff
@@ -78,8 +77,7 @@ public synchronized <PT extends IModelJson> void addJobDefinition(@Nonnull JobDe
 			}
 		}
 
-		Map<String, NavigableMap<Integer, JobDefinition<?>>> newJobsMap = cloneJobsMap();
-		NavigableMap<Integer, JobDefinition<?>> versionMap = newJobsMap.computeIfAbsent(jobDefinitionId, t -> new TreeMap<>());
+		NavigableMap<Integer, JobDefinition<?>> versionMap = myJobDefinitions.computeIfAbsent(jobDefinitionId, t -> new TreeMap<>());
```
Comment (fil512) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobDefinitionRegistry.java:80]: cool.  i knew about sorted maps but not navigable ones...

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobStepExecutor.java [hunk 2]
```diff
@@ -94,6 +95,8 @@ public void executeStep() {
 	private void handleFastTracking(BaseDataSink<PT, IT, OT> theDataSink) {
 		if (theDataSink.getWorkChunkCount() <= 1) {
 			ourLog.debug("Gated job {} step {} produced exactly one chunk:  Triggering a maintenance pass.", myDefinition.getJobDefinitionId(), myCursor.currentStep.getStepId());
+			// wipmb 6.8 either delete fast-tracking, or narrow this call to just this instance and step
```
Comment (fil512) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobStepExecutor.java:98]: fyi fast-tracking is not used in production; it is only used by qa smoke tests.  we could probably find another way to meet qa's needs and remove this completely.

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/maintenance/JobInstanceProcessor.java [hunk 3]
```diff
@@ -133,17 +133,18 @@ private void cleanupInstance(JobInstance theInstance) {
 				break;
 			case CANCELLED:
 				purgeExpiredInstance(theInstance);
+				//wipmb For 6.8 - Are we deliberately not purging chunks for cancelled jobs?  This is a very complicated way to say that.
```
Comment (fil512) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/maintenance/JobInstanceProcessor.java:136]: i would say that is an oversight.  remember CANCELLED used to just be a flag.  the state was added later.  i bet this is an oversight

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/progress/InstanceProgress.java [hunk 4]
```diff
@@ -165,10 +165,6 @@ public void calculateNewStatus() {
 		}
 	}
 
-	public boolean changed() {
```
Comment (fil512) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/progress/InstanceProgress.java:168]: huh.  what did you replace this with??

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/progress/JobInstanceProgressCalculator.java [hunk 5]
```diff
@@ -52,15 +52,12 @@ public void calculateAndStoreInstanceProgress(String theInstanceId) {
 		myJobPersistence.updateInstance(theInstanceId, currentInstance->{
 			instanceProgress.updateInstance(currentInstance);
 
-			if (instanceProgress.changed() || currentInstance.getStatus() == StatusEnum.IN_PROGRESS) {
```
Comment (fil512) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/progress/JobInstanceProgressCalculator.java:55]: much better

**Review by fil512 — APPROVED** [review]
this will probably break some cdr tests

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobDefinitionRegistry.java [hunk 6]
```diff
@@ -33,21 +33,20 @@
...
 import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.stream.Collectors;
 
 public class JobDefinitionRegistry {
 	private static final Logger ourLog = Logs.getBatchTroubleshootingLog();
 
-	// TODO MB is this safe?  Can ue use ConcurrentHashMap instead?
-	private volatile Map<String, NavigableMap<Integer, JobDefinition<?>>> myJobs = new HashMap<>();
+	private final Map<String, NavigableMap<Integer, JobDefinition<?>>> myJobDefinitions = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobDefinitionRegistry.java:49]: It is final now.

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/maintenance/JobInstanceProcessor.java (hunk 3 above)
Comment (michaelabuckley) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/maintenance/JobInstanceProcessor.java:136]: Thanks.  I'll simplify after the release.  Changing this broke a test, and I wanted to double check.

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/progress/InstanceProgress.java (hunk 4 above)
Comment (michaelabuckley) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/progress/InstanceProgress.java:168]: nothing!  the best code.

**Pull request merged**